File: src/dexter/tools/finance/estimates.py
from langchain.tools import tool
from typing import Literal
from pydantic import BaseModel, Field
from dexter.tools.finance.api import call_api

class AnalystEstimatesInput(BaseModel):
    """Input for get_analyst_estimates."""
    ticker: str = Field(..., description="The stock ticker symbol to fetch analyst estimates for. For example, 'AAPL' for Apple.")
    period: Literal['annual', 'quarterly'] = Field(default='annual', description="The period for the estimates, either 'annual' or 'quarterly'.")

@tool(args_schema=AnalystEstimatesInput)
def get_analyst_estimates(
    ticker: str,
    period: Literal['annual', 'quarterly'] = 'annual',
) -> dict:
    """
    Retrieves analyst estimates for a given company ticker, including metrics like estimated EPS.
    Useful for understanding consensus expectations, assessing future growth prospects, and performing valuation analysis.
    """
    params = {
        "ticker": ticker,
        "period": period,
    }
    
    data = call_api("/analyst-estimates/", params)
    return data.get("analyst_estimates", [])

File: src/tools/finance/estimates.ts
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { fmp } from './api.js';
import { formatToolResult } from '../types.js';

export const getAnalystEstimates = new DynamicStructuredTool({
  name: 'get_analyst_estimates',
  description: `Retrieves analyst consensus estimates via FMP, including estimated revenue, EPS, EBITDA, SGA, and net income. Useful for understanding consensus expectations and future growth prospects.`,
  schema: z.object({
    ticker: z
      .string()
      .describe("The stock ticker symbol. For example, 'AAPL' for Apple."),
    period: z
      .enum(['annual', 'quarter'])
      .default('annual')
      .describe("Period for estimates: 'annual' or 'quarter'."),
    limit: z
      .number()
      .default(4)
      .describe('Number of estimate periods to return (default: 4).'),
  }),
  func: async (input) => {
    const ticker = input.ticker.trim().toUpperCase();
    const { data, url } = await fmp.get(`/analyst-estimates/${ticker}`, {
      period: input.period,
      limit: input.limit,
    });
    return formatToolResult(Array.isArray(data) ? data : [], [url]);
  },
});
